org.jacoco.core/src/org/jacoco/core/internal/analysis/filter/FinallyFilter.java

```diff
@@ -0,0 +1,196 @@
+			if (size <= 0) {
+				return;
+			}
+
+			for (final TryCatchBlockNode t : m.tryCatchBlocks) {
+				if (t.handler == finallyBlock.handler) {
+					AbstractInsnNode i = t.start;
+					while (i != t.end) {
+						inside.add(i);
+						i = i.getNext();
+					}
+				}
+			}
+
+			for (final TryCatchBlockNode t : m.tryCatchBlocks) {
+				if (t.handler == finallyBlock.handler) {
+					boolean continues = false;
+					AbstractInsnNode i = t.start;
+
+					while (i != t.end) {
+						switch (i.getType()) {
+						case AbstractInsnNode.FRAME:
+						case AbstractInsnNode.LINE:
+						case AbstractInsnNode.LABEL:
+							break;
+						case AbstractInsnNode.JUMP_INSN:
+							final AbstractInsnNode jumpTarget = next(
+									((JumpInsnNode) i).label);
+							if (!inside.contains(jumpTarget)) {
+								merge(size, e, jumpTarget);
+							}
+							continues = i.getOpcode() != Opcodes.GOTO;
+							break;
+						default:
+							switch (i.getOpcode()) {
+								case Opcodes.IRETURN:
+								case Opcodes.LRETURN:
+								case Opcodes.FRETURN:
+								case Opcodes.DRETURN:
+								case Opcodes.ARETURN:
+								case Opcodes.RETURN:
+								case Opcodes.ATHROW:
+									continues = false;
+									break;
+								default:
+									continues = true;
+									break;
+							}
+							break;
+						}
+						i = i.getNext();
+					}
+
+					i = next(i);
+					if (continues && !inside.contains(i)) {
+						merge(size, e, i);
+					}
+				}
+
+				if (t != finallyBlock && t.start == finallyBlock.start
+						&& t.end == finallyBlock.end) {
+					final AbstractInsnNode i = next(next(t.handler));
+					if (!inside.contains(i)) {
+						// javac's empty catch - merge after ASTORE
+						merge(size, e, i);
+					}
+				}
+			}
+		}
+
+		private void merge(final int size, AbstractInsnNode e,
+				AbstractInsnNode n) {
+			if (!isSame(size, e, n)) {
+				return;
+			}
+			output.ignore(e, e);
+			e = next(e);
+			for (int i = 0; i < size; i++) {
+				output.merge(e, n);
+				e = next(e);
+				n = next(n);
+			}
+			output.ignore(e, next(e));
+			if (n != null && n.getOpcode() == Opcodes.GOTO) {
+				output.ignore(n, n);
+			}
+		}
+
+	}
+
+	private static boolean isSame(final int size, AbstractInsnNode e,
+			AbstractInsnNode n) {
+		e = next(e);
+		for (int i = 0; i < size; i++) {
+			if (n == null || e.getOpcode() != n.getOpcode()) {
+				return false;
+			}
+			e = next(e);
+			n = next(n);
+		}
+		return true;
+	}
+
+	/**
+	 * @return number of instructions inside given finally handler
+	 */
+	private static int size(AbstractInsnNode i) {
+		if (Opcodes.ASTORE != i.getOpcode()) {
+			// when always completes abruptly
+			return 0;
+		}
+		final int var = ((VarInsnNode) i).var;
+		int size = -1;
+		do {
+			size++;
+			i = next(i);
+			if (i == null) {
+				// when always completes abruptly
+				return 0;
+			}
+		} while (!(Opcodes.ALOAD == i.getOpcode()
+				&& var == ((VarInsnNode) i).var));
+		i = next(i);
+		if (Opcodes.ATHROW != i.getOpcode()) {
```

Comment:
@Godin Potential NPE? `i` could be `null` here -- but only for invalid class files. So probably not an issue.